docs/proposals/kubelet-qos.md

```diff
@@ -0,0 +1,125 @@
+
+## Implementation
+
+Implement QoS policy
+- Add a function in kubelet.go that decides whether a pod is best-effort or top-tier
+- For now, return based on whether container limits are specified, but eventually this can be refactored based on a user specified value in the pod spec
+
+Change memory overcommit mode
+- Right now overcommit mode is off on the machines we set up, so if there isn’t enough memory malloc will return null. This prevents QoS, because best-effort pods won’t be evicted. Instead, when there isn’t enough memory, and top-tier pods call malloc, they may not get the memory they want. We want top-tier pods to get the memory they request, and force out best-effort pods.
+- Change the memory overcommit mode to 1, so that using excess memory starts the OOM killer.
+
+Pod admission
+- Modify predicates.CheckPodsExceedingCapacity (in scheduler/algorithm/predicates, which is used by both the Kubelet and the Scheduler)
+- If the pod is best-effort, we admit it automatically
+- Otherwise, admit the top-tier pod if there’s space based on limits of the other top-tier pods
+
+Pod configuration
+- We’re focusing on Docker in this implementation (not RKT)
+- OOM scores
+  - Note that the OOM score of a process is 10 times the % of memory the process consumes, adjusted by OOM_SCORE_ADJ, barring exceptions (e.g. process is launched by different users). Processes with higher OOM scores are killed.
+  - The base OOM score is between 0 and 1000, so if process A’s OOM_SCORE_ADJ - process B’s OOM_SCORE_ADJ is over a 1000, then process A will die.
+- Best-effort
+  - In the initial implementation, ignore CPU limits, set CPU shares to 2. In the final implementation, create a “best-effort” cgroup that contains all the best-effort pods. The best-effort cgroup will get 2 CPU shares. Within the cgroup, each container will be in a cgroup that gets CPU shares proportional to the amount of CPU it request for (or 2 shares if no limit was specified).
+  - Set memory limits as specified by the user
+  - Set OOM_SCORE_ADJ to 501
+- Top-tier
+  - Set CPU shares and memory requirements as we do right now
+  - Set OOM_SCORE_ADJ to -501
+- Special pods
+  - The Pod infrastructure container gets an OOM_SCORE_ADJ of -701
+  - Kubelet, Kube-Proxy, etc will get an even lower OOM_SCORE_ADJ
+
+Ensuring resource guarantees
+- After being configured suitably, the kernel will handle resource guarantees as specified in QoS Functionality
+
+Setting the OOM score
+- To set OOM_SCORE_ADJ of a container, we loop through all processes in the container, and set OOM_SCORE_ADJ

```

Comment:
@gmarek Actually, you're right, the race is probably not *that* rare. I mentioned one of the problems that the race could cause towards the end of my proposal. If you (or anyone else) have any suggestions on how to mitigate the race, please let me know! :)